src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/measure/ActionMeter.kt

```diff
@@ -76,50 +97,80 @@ class ActionMeter(
         observation: (T) -> JsonObject?
     ): T {
         val start = clock.instant()
+        var actionResult: ActionResult = ActionResult.OK
+        var result: T? = null
+
         try {
-            val result = action()
-            output.write(
-                ActionMetric.Builder(
-                    label = key.label,
-                    result = ActionResult.OK,
-                    start = start,
-                    duration = Duration.between(start, clock.instant())
-                )
-                    .virtualUser(virtualUser)
-                    .observation(observation(result))
-                    .drilldown(w3cPerformanceTimeline.record())
-                    .build()
-            )
+            result = action()
             return result
         } catch (e: Exception) {
-            val result = if (e.representsInterrupt()) {
+            actionResult = if (e.representsInterrupt()) {
                 ActionResult.INTERRUPTED
             } else {
                 ActionResult.ERROR
             }
-            output.write(
-                ActionMetric.Builder(
-                    label = key.label,
-                    result = result,
-                    start = start,
-                    duration = Duration.between(start, clock.instant())
-                )
-                    .virtualUser(virtualUser)
+            throw Exception("Action '${key.label}' $actionResult", e)
+        } finally {
+            val metricBuilder = ActionMetric.Builder(
+                label = key.label,
+                result = actionResult,
+                start = start,
+                duration = Duration.between(start, clock.instant())
+            )
+                .virtualUser(virtualUser)
+
+            result?.let { metricBuilder.observation(observation(result)) }
+
+            var metric = metricBuilder.build()
+
+            if (drilldownCondition.test(metric)) {
+                metric = ActionMetric.Builder(metric)
                     .drilldown(w3cPerformanceTimeline.record())
                     .build()
-            )
-            throw Exception("Action '${key.label}' $result", e)
+            }
+            output.write(metric)
         }
     }
 
+    @Deprecated( message = "Use builder instead")
```

Comment:
Pro-tip, you can use [`ReplaceWith`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-replace-with/) to provide a top-notch migration experience.